pkg/policy/mapstate.go

```diff
@@ -388,11 +502,29 @@ func protocolsMatch(a, b Key) bool {
 	return a.Nexthdr == 0 || b.Nexthdr == 0 || a.Nexthdr == b.Nexthdr
 }
 
+// RevertChanges undoes changes to 'keys' as indicated by 'changes.adds' and 'changes.old' collected via
+// denyPreferredInsertWithChanges().
+func (keys MapState) RevertChanges(changes ChangeState) {
+	for k := range changes.Adds {
+		delete(keys, k)
+	}
+	// 'old' contains all the original values of both modified and deleted entries
+	for k, v := range changes.Old {
+		keys[k] = v
+	}
+}
+
 // denyPreferredInsertWithChanges contains the most important business logic for policy insertions. It inserts
 // a key and entry into the map by giving preference to deny entries, and L3-only deny entries over L3-L4 allows.
 // Incremental changes performed are recorded in 'adds' and 'deletes', if not nil.
 // See https://docs.google.com/spreadsheets/d/1WANIoZGB48nryylQjjOw6lKjI80eVgPShrdMTMalLEw#gid=2109052536 for details
-func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, adds, deletes Keys, identities Identities) {
+func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStateEntry, identities Identities, features policyFeatures, changes ChangeState) {
+	// Skip deny rules processing if the policy in this direction has no deny rules
+	if !features.contains(denyRules) {
```

Comment:
This skip of the deny logic when no denies are present is purely an optimization made possible by tracking of policy features that was added mainly to cut the additional cost of the auth-logic when authentication is not used in the policy. My recollection is that on the the ConsumeMapChanges code path the direction specific features would have been harder to use, as the collected map changes can contain entries in both directions. In that case we end up doing some unnecessary deny/auth logic for the direction where those features were not actually present.

`authPreferredInsert` call is not skipped in any case, but if deny feature is indicated it will be done at the bottom of `denyPreferredInsert` after scanning through all the mapstate entries.